lib/spack/llnl/util/filesystem.py

```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
         os.unlink(path)
 
 
+class GenericVisitor(object):
+    def __init__(self, callable):
+        pass
+
+    def before_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def before_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def visit_file(self, root, rel_path, depth):
+        pass
+
+    def visit_symlinked_file(self, root, rel_path, depth):
+        pass
+
+
+class CallingVisitor(GenericVisitor):
+    def __init__(self, callable):
+        self.caller = callable
+
+    def before_visit_dir(self, root, rel_path, depth):
+        return True
+
+    def after_visit_dir(self, root, rel_path, depth):
+        self.caller(os.path.join(root, rel_path))
+
+    def before_visit_symlinked_dir(self, root, rel_path, depth):
+        return True
+
+    def after_visit_symlinked_dir(self, root, rel_path, depth):
+        self.caller(os.path.join(root, rel_path))
+
+    def visit_file(self, root, rel_path, depth):
+        self.caller(os.path.join(root, rel_path))
+
+    def visit_symlinked_file(self, root, rel_path, depth):
+        self.caller(os.path.join(root, rel_path))
+
+
+class MakeWinWriteableVisitor(GenericVisitor):
```

Comment:
(naming) While this visitor is only used for this purpose, it probably makes sense to name this more based on it's behavior WRT how it decides to visit files/dirs/symlinks.